app/bundles/CampaignBundle/Executioner/RealTimeExecutioner.php

```diff
@@ -240,14 +249,36 @@ private function evaluateDecisionForContact(Event $event, $passthrough = null, $
         }
 
         // If channels do not match up at all (not even fuzzy logic i.e. page vs page.redirect), there's no need to go further
-        if ($channel && $event->getChannel() && $channel && strpos($channel, $event->getChannel()) === false) {
+        if ($channel && $event->getChannel() && strpos($channel, $event->getChannel()) === false) {
             throw new DecisionNotApplicableException("Channels, $channel and {$event->getChannel()}, do not match.");
         }
 
         if ($channel && $channelId && $event->getChannelId() && $channelId !== $event->getChannelId()) {
             throw new DecisionNotApplicableException("Channel IDs, $channelId and {$event->getChannelId()}, do not match for $channel.");
         }
 
+        // Check if parent taken path is the path of this event, otherwise exit
+        $parentEvent = $event->getParent();
+        if ($parentEvent !== null && $event->getDecisionPath() !== null) {
+            $rotation    = $this->leadRepository->getContactRotations([$this->contact->getId()], $event->getCampaign()->getId());
+            $log = $parentEvent->getLogByContactAndRotation($this->contact, $rotation);
+
+            if ($log === null) {
+                throw new DecisionNotApplicableException("Parent {$parentEvent->getId()} has not been fired, event {$event->getId()} should not be fired.");
+            }
+
+            $pathTaken   = (int)$log->getNonActionPathTaken();
+            $isScheduled = (int)$log->getIsScheduled();
+
+            if ($isScheduled === 1) {
+                throw new DecisionNotApplicableException("Parent {$parentEvent->getId()} has not been fired, event {$event->getId()} should not be fired.");
```

Comment:
is_scheduled will be 1 if the event was scheduled and later executed, so you need to also check `$log->getTriggerDate() < now()` or similar. In other words "is_scheduled" doesn't change to a 0 when the event happens.

Reply:
I have made a lot of test and when event is triggered `is_scheduled` column is set to 0.

Reply:
Ex: third and fourth are scheduled events
![screenshot_2018-11-29 127 0 0 1 127 0 0 1 mautic-wmk campaign_lead_event_log phpmyadmin 4 7 0](https://user-images.githubusercontent.com/27768270/49232597-8abcd100-f3f4-11e8-9845-2ae4dea30e2b.png)

Reply:
How is the even triggered though? Real time, or campaign executioner? First attempt or second? There are different kinds of scheduled events. There are events that are intentionally delayed, then there are events that fail (due to integration issue or similar), and *become* scheduled. 

For example, I have over 300 million campaign_lead_event_log entries in production with is_scheduled set to 1 that were triggered in the past. They are not counted as events to be triggered by the campaign executioner.

Unless... something has recently changed drastically in staging, and there's a massive migration coming my way... 😨

Reply:
You were right about this... I was thinking of a scenario of a non-scheduled action which can become scheduled when there's a failure... Core doesn't behave the same way as it did last I looked at this.